README
NAME
    Geo::Google - Perform geographical queries using Google Maps

SYNOPSIS
      use strict;
      use Data::Dumper;
      use Geo::Google;

      #My office
      my $gonda_addr = '695 Charles E Young Dr S, Westwood, CA 90024';
      #Stan's Donuts
      my $stans_addr = '10948 Weyburn Ave, Westwood, CA 90024';

      #Instantiate a new Geo::Google object.
      my $geo = Geo::Google->new();

      #Create Geo::Google::Location objects.  These contain
      #latitude/longitude coordinates, along with a few other details
      #about the locus.
      my ( $gonda ) = $geo->location( address => $gonda_addr );
      my ( $stans ) = $geo->location( address => $stans_addr );
      print $gonda->latitude, " / ", $gonda->longitude, "\n";
      print $stans->latitude, " / ", $stans->longitude, "\n";

      #Create a Geo::Google::Path object.
      my ( $donut_path ) = $geo->path($gonda,$stans);

      #A path contains a series of Geo::Google::Segment objects with
      #text labels representing turn-by-turn driving directions between
      #the two loci.
      my @segments = $donut_path->segments();

      #This is the human-readable directions for the first leg of the
      #journey.
      print $segments[0]->text(),"\n";

      #Geo::Google::Segment objects contain a series of
      #Geo::Google::Location objects -- one for each time the segment
      #deviates from a straight line to the end of the segment.
      my @points = $segments[1]->points;
      print $points[0]->latitude, " / ", $points[0]->longitude, "\n";

      #Now how about some coffee nearby?
      my @coffee = $geo->near($stans,'coffee');
      #Too many.  How about some Coffee Bean & Tea Leaf?
      @coffee = grep { $_->title =~ /Coffee.*?Bean/i } @coffee;

      #Still too many.  Let's find the closest with a little trig and
      #a Schwartzian transform
      my ( $coffee ) = map { $_->[1] }
                       sort { $a->[0] <=> $b->[0] }
                       map { [ sqrt(
                                ($_->longitude - $stans->longitude)**2
                                  +
                                ($_->latitude - $stans->latitude)**2
                               ), $_ ] } @coffee;

DESCRIPTION
    Geo::Google provides access to the map data used by the popular Google
    Maps web application.

  WHAT IS PROVIDED
    Conversion of a street address to a 2D Cartesian point
    (latitude/longitude)
    Conversion of a pair of points to a multi-segmented path of driving
    directions between the two points.
    Querying Google's "Local Search" given a point and one or more query
    terms.

  WHAT IS NOT PROVIDED
    Documentation of the Google Maps map data XML format
    Documentation of the Google Maps web application API
    Functionality to create your own Google Maps web page.

AUTHOR
    Allen Day <allenday@ucla.edu>

    Copyright (c) 2004-2005 Allen Day. All rights reserved. This program is
    free software; you can redistribute it and/or modify it under the same
    terms as Perl itself.

BUGS / TODO
    Report documentation and software bugs to the author, or better yet,
    send a patch. Known bugs/issues:

    Polyline decoding needs to be cleaned up.
    Lack of documentation.

SEE ALSO
      http://maps.google.com
      http://www.google.com/apis/maps/
      http://libgmail.sourceforge.net/googlemaps.html


Changes
Revision history for Perl extension Geo::Google.

0.01  Mon May 23 22:05:45 2005
	- original version; created by h2xs 1.23 with options
		-Xn Geo::Google
0.03  Sun Jan 7 10:21:00 2007
        - Google switched their Maps server response from XML
            to JSON.  Changed the Geo::Google response parsing
            from XML to google's JSON data structure.
        - changed the toString() functions of these objects to
            toXML() functions:
               Geo::Google::Location
               Geo::Google::Segment
               Geo::Google::Path
	- Added a toJSON() function to these objects:
               Geo::Google::Location
               Geo::Google::Path
        - Added private function _JSONRenderSkeleton() to support
            the new toJSON() functions of location and path objects
        - Simplified control loops for directions segment parsing
        - Added support for waypoints in directions queries and
            Geo::Google::Path objects
        - Updated the test script
        - Added the levels() and panel() methods of the 
	    Geo::Google::Path object to allow JSON rendering that 
	    should be interchangeable with Google Maps' JSON response
        - Added a locations() method of the Geo::Google::Path
            object, allowing you to access any array of all 
	    locations used in the directions query after it has 
	    been executed
	- Expanded error handling to include more checks during the 
	    query process, added new error messages, and re-wrote 
	    some existing error messages for clarity
	- Added detailed test failure reporting to t/Geo-Google.t
20070226:0213:MAT 
	Partial fix of a bug in extraction and correlation of 
	driving directions steps.
20070227:0516:MAT 
	-Completed fix of Path segments bug reported by
		Thomas Delaet and Pieter-Jan Speelmans
	-Added code to automatically accept suggested changes
		to google maps queries in the Geo::Google->path() 
		function.
	-Adjusted path object to include start, stop and all 
		waypoints as segments
	-Revised test script to reflect new path data structure
20070301:0458:MAT
	-Cleaned up suggestion acceptor in Geo::Google->path()
		-Based parsing on a style tag instead of English text 
	-Replaced parsing for html segment descriptions with 
		a description constructor for waypoints only 
20070302:1422:MAT
	-Removed waypoint html segment description contstructor.
		Replaced it with the official Google title of the 
		location from from the Geo::Google::Location object
		for the waypoint.
20070304:1751:MAT
	-Fixed parsing bug that caused directions with no waypoints
		to miss all directions steps, returning only the 
		start and stop points.

		Special thanks to Pieter Agten <pi8 at pi.be>
		for zeroing in on the problem line and sending a 
		patch.
	-Added 14 new tests to verify that directions with no 
		waypoints (from: to:) parse correctly.
20070308:0626:MAT
	-re-tooled driving directions parsing and adjusted structure 
		of the path object slightly to comply with Google's 
		latest changes to the JSON data.  Added a second
		polyline (because google switched to one polyline per 
		subpath)
		-Thanks to Pieter Agten for spotting this and sending
		 in a bug report
0.04-rc1 : 9 April 2007 : MAT
    -Fixed broken distance and time parsing
    -Re-wrote and added extensive commenting to the polyline decoding
    	- now it works on 64-bit perl as well
    -Adjusted HTML segment to polyline points synchronization loops
     to account for the new polyline decoding algorithm
0.04-rc2 : MAT : adjusted for slightly different point lat/longs in test script
0.04-rc3 : MAT
  - This version requires perl JSON module 1.11 patched with Allen Day's
      JSON/Parser.pm patch.  Recently added JSON from Google Maps broke 
      the JSON parser.  The Allen Day patch makes it more flexible for
      quote handling.
  - Adjusted the _obj2location() function because Google changed their
      JSON data structure for locations (address lines specifically).
0.04-rc4: MAT
  - Rewrote _encode() for 64 bit CPU compatibility
  - Wrote _encode_word() function used by new version of _encode()
  - Fixed Geo::Google::Path->toXML() function (have not tested for KML file
    compatibility).
  - Added static unit tests for:
    _encode_word()
    _encode()
    _decode_word()
    _decode()
    _decode() to _encode() and comparison of output against input
  - Added a dynamic unit test that tests both _decode() and _encode() against
    the actual Google polyline string
  - Adjusted the correlation algorithm for HTML extracted directions segments and
    points extracted from the polyline strings.  We now use search locations for
    points wherever possible.
    - To make this possible, I adjusted precision of lat/long coords for 
      Geo::Google::Locations that are used as path search parameters.  They come
      from Google with millionth degree precision, but polyline points only have
      hundred-thousandth degree precision.
    - Points to HTML correlation is spot on for point-to-point directions, but off
      for point-to-point-to-point directions (waypoints).  Will fix this when I
      restructure the API for Path->Subpath->Segment directions structure.
  - Adjusted directions parsing to accommodate Google's naming convention change
    for driving segments.  It was 'dirsegment_\d+_\d+', but now it's 
    'panel_\d+_\d+'.
0.04-rc5 (MAT)
  - Modified the location() constructor method of of the Geo::Google object to
    allow it to return Geo::Gooogle::Location objects when Google can't resolve
    the address you provide but suggests alternate addresses that it can
    resolve.  It returns these alternate address in the same array of
    Geo::Google::Location objects that it normally returns.  They are ordered 
    in the same order Google presents them.
  - Added unit tests for handling of suggested alternate addresses in the
    location() constructor method of the Geo::Google object.
0.04-rc5 (MAT)

Makefile.PL
use ExtUtils::MakeMaker;
# See lib/ExtUtils/MakeMaker.pm for details of how to influence
# the contents of the Makefile that is written.
WriteMakefile(
    NAME              => 'Geo::Google',
    VERSION_FROM      => 'lib/Geo/Google.pm', # finds $VERSION
    PREREQ_PM         => {
Digest::MD5     => 0.0,
HTML::Entities  => 0.0,
JSON		=> 0.0,
LWP::Simple     => 0.0,
URI::Escape     => 0.0,
                         }, # e.g., Module::Name => 1.1
    ($] >= 5.005 ?     ## Add these new keywords supported since 5.005
      (ABSTRACT_FROM  => 'lib/Geo/Google.pm', # retrieve abstract from module
       AUTHOR         => 'Allen Day E<lt>allenday@ucla.eduE<gt> ' .
				'(Initial through 0.2)\n' .
				'Revision 0.3 maintainer:\n  ' .
				'Michael Trowbridge E<lt>' .
				'michael.a.trowbridge@gmail.comE<gt>') : ()),
);


lib/Geo/Google/Segment.pm
package Geo::Google::Segment;
use strict;
use warnings;
use Data::Dumper;
use URI::Escape;
our $VERSION = '0.03';

use constant FMT => <<_FMT_;
<segment distance="%s" id="%s" pointIndex="%s" time="%s">%s</segment>
_FMT_

sub new {
  my $class = shift;
  my %arg = @_;
  my $self = bless \%arg, $class;
}

sub from       { return shift->{'from'} }
sub distance   { return shift->{'distance'} }
sub id         { return shift->{'id'} }
sub pointIndex { return shift->{'pointIndex'} }
sub points     { my $self = shift; return $self->{'points'} ? @{ $self->{'points'} } : () }
sub html       { return shift->{'html'} }
sub text       { my $self = shift; $_ = $self->html(); s#<script.+?</script##g; s#<.+?># #g; s#\s+# #g; s# ([,:])#$1#g; s#^\s*(.+?)\s*$#$1#; $_ }
sub time       { return shift->{'time'} }
sub to         { return shift->{'to'} }

sub toXML {
  my $self = shift;
  return sprintf( FMT,
    $self->distance(),
    $self->id(),
    $self->pointIndex(),
    $self->time(),
    $self->text(),
  );
}

1;
__END__
# Below is stub documentation for your module. You'd better edit it!

=head1 NAME

Geo::Google::Segment - An annotated segment of a path

=head1 SYNOPSIS

  use Geo::Google::Segment;
  # you shouldn't need to construct these yourself,
  # have a Geo::Google object do it for you.

=head1 DESCRIPTION

Google Maps is able to serve up directions between two points.  Directions
consist of two types of components:

  1. a series of points along a "polyline".
  2. a series of annotations, each of which applies to a contiguous
  range of points.

In the Geo::Google object model, directions are available by calling path()
on a Geo::Google instance.  The return value is a Geo::Google::Path object,
which is a composite of Geo::Google::Segment objects, which are in turn
composites of Geo::Google::Location objects.

In Google Earth, they're used as XML inside a .KML file or xml 
server response.  Individual segments are used in Google Maps as 
HTML with JavaScript inside a JSON response from the server.  
The JSON structure used by Google Maps doesn't actually contain 
segments.  They are implied by the HTML in the "panel" portion of 
the response and activate JavaScript that re-renders the map window 
after extracting coordinates from the polyline object. 


=head1 OBJECT METHODS

Geo::Google::Segment objects provide the following accessor methods

 Method      Description
 ------      -----------
 from        a Geo::Google::Location at the beginning of the segment.
 distance    length of the segment, in variable, human friendly units.
 id          a unique identifier for this point.
 pointIndex  the offset, in 0-based points, at which the segment begins
             relative to the start of the path.
 points      a list of Geo::Google::Location points along the segment.
 html        a description of the segment, typically some human-readable
             description of this leg of a path, e.g "make a u-turn at
             McLaughlin Ave".
 text        a scrubbed version of the html() function's output
 time        a time estimate, in variable, human-friendly units for how long
             the segment will take to travel by automobile.
 to          a Geo::Google::Location at the end of the segment.
 toXML       a method that renders the segment in XML that could be used as part of a Google Earth KML file.

=head1 SEE ALSO

L<Geo::Google>

=head1 AUTHOR

Allen Day E<lt>allenday@ucla.eduE<gt>, Michael Trowbridge 
E<lt>michael.a.trowbridge@gmail.comE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (c) 2004-2007 Allen Day.  All rights
reserved. This program is free software; you can redistribute it
and/or modify it under the same terms as Perl itself.

=cut


lib/Geo/Google/Location.pm
package Geo::Google::Location;
use strict;
use warnings;
use Data::Dumper;
use URI::Escape;
use JSON;
use Geo::Google;
our $VERSION = '0.03';

use constant FMT => <<_FMT_;
<location infoStyle="%s" id="%s">
  <point lat="%s" lng="%s"/>
  <icon image="%s" class="local"/>
  <info>
    <address>
      %s
    </address>
  </info>
</location>
_FMT_

#      $loc->{'latitude'} = $lat;
#      $loc->{'longitude'} = $lng;
#      $loc->{'lines'} = [@lines];
#      $loc->{'id'} = $id;
#      $loc->{'icon'} = $icon;
#      $loc->{'infostyle'} = $infoStyle;

sub new {
  my $class = shift;
  my %arg = @_;
  my $self = bless \%arg, $class;
}

sub icon      { return shift->{'icon'} }
sub id        { return shift->{'id'} }
sub infostyle { return shift->{'infostyle'} }
sub latitude  { return shift->{'latitude'} }
sub lines     { my $self = shift; return $self->{'lines'} ? @{ $self->{'lines'} } : () }
sub longitude { return shift->{'longitude'} }
sub title     { return shift->{'title'} }

sub toXML {
  my $self = shift;
  return sprintf( FMT,
    $self->infostyle(),
    $self->id(),
    $self->latitude(),
    $self->longitude(),
    $self->icon(),
    join('',map {"<line>$_</line>"} $self->lines() ),
  );
}

sub toJSON {
	my $self = shift;
	my $json = new JSON (barekey => 1);

	# Construct the shell of a new perl data structure that we
	# can pass off to the JSON module for rendering to a string
	my $preJSON = Geo::Google::_JSONrenderSkeleton();

	# Fill the perl data structure with information from this
	# location
	$preJSON->{"form"}->{"l"}->{"near"} = $self->title();
	$preJSON->{"form"}->{"q"}->{"q"} = $self->title();
	$preJSON->{"form"}->{"d"}->{"daddr"} = $self->title();
	$preJSON->{"form"}->{"selected"} = "q";
	push( @{$preJSON->{"overlays"}->{"markers"}}, {
		"icon" => "addr",
		"laddr" => $self->title(),
		"svaddr" => $self->title(),
		"lat" => $self->latitude(),
		"lng" => $self->longitude(),
		"id" => "addr",
		"adr" => 1,
		"image" => "/mapfiles/arrow.png",
		"elms" => [(3, 2, 6)],
		"dtlsUrl" => "/maps?v=1&q=" . uri_escape($self->title())
			. "&ie=UTF8&hl=en&latlng=&ei=",
		"infoWindow" => {
				"basics" => join("<BR />", $self->lines()),
				"type" => "html",
				"_" => 0
				}
		});
	$preJSON->{"printheader"} = $self->title();
	$preJSON->{"viewport"}->{"span"}->{"lat"} = "0.089989";
	$preJSON->{"viewport"}->{"span"}->{"lng"} = "0.107881";
	$preJSON->{"viewport"}->{"center"}->{"lat"} = $self->latitude();
	$preJSON->{"viewport"}->{"center"}->{"lng"} = $self->longitude();
	$preJSON->{"url"} = "/maps?v=1&q=" . 
				uri_escape($self->title()) . "&ie=UTF8";
	$preJSON->{"title"} = $self->title();

	# Render the data structure to a JSON string and return it
	return $json->objToJson($preJSON);
}

1;
__END__
# Below is stub documentation for your module. You'd better edit it!

=head1 NAME

Geo::Google::Location - A geographical point

=head1 SYNOPSIS

  use Geo::Google::Location;
  # you shouldn't need to construct these yourself,
  # have a Geo::Google object do it for you. 

=head1 DESCRIPTION

=head1 OBJECT METHODS

Geo::Google::Location objects provide the following accessor methods

 Method      Description
 ------      -----------
 icon        an icon to use when drawing this point.
 id          a unique identifier for this point.
 infostyle   determines how a pop-up info window callout for a plotted point is displayed in Google Maps or Google Earth
 latitude    latitude of the point, to hundred-thousandth degree precision.
 lines       a few lines describing the point, useful as a label
 longitude   longitude of the point, to hundred-thousandth degree precision.
 title       a concise description of the point.
 toXML       a method that renders the point in XML format for use in 
             Google Earth KML files.
 toJSON	     a method that renders the point in JSON format for use
             with the Google Maps API.

=head1 SEE ALSO

L<Geo::Google>

=head1 AUTHOR

Allen Day E<lt>allenday@ucla.eduE<gt>, Michael Trowbridge 
E<lt>michael.a.trowbridge@gmail.comE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (c) 2004-2007 Allen Day.  All rights
reserved. This program is free software; you can redistribute it
and/or modify it under the same terms as Perl itself.

=cut


lib/Geo/Google/Path.pm
package Geo::Google::Path;
use strict;
use warnings;
use Data::Dumper;
use URI::Escape;
use Geo::Google;
our $VERSION = '0.03';

use constant FMT => <<_FMT_;
<segments distance="%s" time="%s">
%s</segments>
_FMT_

#<segments distance="0.6&#160;mi" meters="865" seconds="56" time="56 secs">
#  <segment distance="0.4&#160;mi" id="seg0" meters="593" pointIndex="0" seconds="38" time="38 secs">Head <b>southwest</b> from <b>Venice Blvd</b></segment>
#  <segment distance="0.2&#160;mi" id="seg1" meters="272" pointIndex="6" seconds="18" time="18 secs">Make a <b>U-turn</b> at <b>Venice Blvd</b></segment>
#</segments>

sub new {
  my $class = shift;
  my %arg = @_;
  my $self = bless \%arg, $class;
}

sub distance { return shift->{'distance'} }
sub polyline { return shift->{'polyline'} }
sub segments { my $self = shift; return $self->{'segments'} ? @{ $self->{'segments'} } : () }
sub time     { return shift->{'time'} }
sub locations { my $self = shift; return $self->{'locations'} ? @{ $self->{'locations'} } : () }
sub panel    { return shift->{'panel'} }
sub levels   { return shift->{'levels'} }

sub toXML {
  my $self = shift;
  my $content = join "", map { $_->toXML } @{ $self->segments };
  return sprintf( FMT,
    $self->distance(),
    $self->time(),
    $content,
  );
}

sub toJSON {
	my $self = shift;
	my $json = new JSON (barekey => 1);

	# Construct the shell of a new perl data structure that we
	# can pass off to the JSON module for rendering to a string
	my $preJSON = Geo::Google::_JSONrenderSkeleton();

	# Fill the perl data structure with information from this
	# location
	my @locations = $self->locations();
	$preJSON->{"form"}->{"l"}->{"near"} = 
				$locations[$#locations]->title();
	$preJSON->{"form"}->{"q"}->{"q"} = "from:" . 
					$locations[0]->title();
	for (my $i=1; $i<=$#locations; $i++) {
		$preJSON->{"form"}->{"q"}->{"q"} .= " to:" .
			$locations[$i]->title();
	}
	$preJSON->{"form"}->{"d"}->{"saddr"} = $locations[0]->title();
	$preJSON->{"form"}->{"d"}->{"daddr"} = $locations[1]->title();
	for (my $i=2; $i<=$#locations; $i++) {
		$preJSON->{"form"}->{"d"}->{"daddr"} .= " to:" .
			$locations[$i]->title();
	}
	$preJSON->{"form"}->{"selected"} = "q";
	$preJSON->{"panelStyle"} = "";
	# Generate the markers
	for (my $i=0; $i<=$#locations; $i++) {
		my $image = "/mapfiles/dd-pause.png";
		my $id = "pause";
		if ($i == 0 ) {
			$image = "/mapfiles/dd-start.png";
			$id = "start";
		}
		elsif ($i == $#locations) {
			$image = "/mapfiles/dd-stop.png";
			$id = "stop";
		}
		push( @{$preJSON->{"overlays"}->{"markers"}}, {
			"laddr" => "",
			"svaddr" => "",
			"lat" => $locations[$i]->latitude(),
			"lng" => $locations[$i]->longitude(),
			"id" => $id,
			"image" => $image,
			"elms" => [()],
			"llcid" => "",
			"fid" => "",
			"cid" => "",
			"sig" => "",
			"infoWindow" => {
				"type" => "map",
				"_" => 0
				}
			});
	}
	$preJSON->{"overlays"}->{"polylines"}->[0]->{"levels"} =
		$self->levels();
	$preJSON->{"overlays"}->{"polylines"}->[0]->{"points"} =
		$self->polyline();	
	$preJSON->{"overlays"}->{"polylines"}->[0]->{"outline"} = undef;
	$preJSON->{"overlays"}->{"polylines"}->[0]->{"opacity"} = undef;
	$preJSON->{"overlays"}->{"polylines"}->[0]->{"color"} = undef;
	$preJSON->{"overlays"}->{"polylines"}->[0]->{"weight"} = undef;
	$preJSON->{"overlays"}->{"polylines"}->[0]->{"id"} = "d";
	$preJSON->{"overlays"}->{"polylines"}->[0]->{"numLevels"} = 4;
	$preJSON->{"overlays"}->{"polylines"}->[0]->{"zoomFactor"} = 16;
	$preJSON->{"printheader"} = sprintf("Directions from %s to %s"
					. "<BR />Travel time: %s", 
					$locations[0]->title(),
					$locations[$#locations]->title(),
					$self->time());
	$preJSON->{"modules"} = [(undef, "multiroute")];
	$preJSON->{"panelResizeState"} = "open";
	# step through all points in the directions and determine
	# the maximum and minimum lat/longs
	my $maxlat = -90;
	my $minlat = 90;
	my $maxlng = -180;
	my $minlng = 180;
	foreach my $segment ($self->segments()) {
		foreach my $point ($segment->points()) {
			if ($point->latitude() > $maxlat) 
				{ $maxlat = $point->latitude(); }
			if ($point->latitude() < $minlat) 
				{ $minlat = $point->latitude(); }
			if ($point->longitude() > $maxlng) 
				{ $maxlng = $point->longitude(); }
			if ($point->longitude() < $minlng) 
				{ $minlng = $point->longitude(); }
		}
	}
	$preJSON->{"viewport"}->{"span"}->{"lat"} = sprintf("%.6f",
		( ($maxlat-$minlat) * 1.1) );
	$preJSON->{"viewport"}->{"span"}->{"lng"} = sprintf("%.6f", 
		( ($maxlng-$minlng) * 1.1) );		
	$preJSON->{"viewport"}->{"center"}->{"lat"} = sprintf("%.6f", 
		( ($maxlat+$minlat) / 2) );
	$preJSON->{"viewport"}->{"center"}->{"lng"} = sprintf("%.6f",
		( ($maxlng+$minlng) / 2) );
	$preJSON->{"viewport"}->{"mapType"} = undef;
	$preJSON->{"url"} = "/maps?v=1&q=" . 
				uri_escape(sprintf("from:%s to:%s",
					$locations[0]->title(), 
					$preJSON->{"form"}->{"d"}->{"daddr"}))
				. "&ie=UTF8";
	$preJSON->{"title"} = sprintf("From:%s to:%s", 
				$locations[0]->title(),
				$preJSON->{"form"}->{"d"}->{"daddr"});
	$preJSON->{"panel"} = $self->panel();

	# Render the data structure to a JSON string and return it
	return $json->objToJson($preJSON);
}

1;
__END__
 Below is stub documentation for your module. You'd better edit it!

=head1 NAME

Geo::Google::Path - A path, by automobile, between two loci.

=head1 SYNOPSIS

  use Geo::Google::Path;
  # you shouldn't need to construct these yourself,
  # have a Geo::Google object do it for you.

=head1 DESCRIPTION

Google Maps is able to serve up directions between two points.  Directions
consist of two types of components:

  1. a series of points along a "polyline".
  2. a series of annotations, each of which applies to a contiguous
  range of points.

In the Geo::Google object model, directions are available by calling path()
on a Geo::Google instance.  The return value is a Geo::Google::Path object,
which is a composite of Geo::Google::Segment objects, which are in turn
composites of Geo::Google::Location objects.

=head1 OBJECT METHODS

Geo::Google::Path objects provide the following accessor methods

 Method      Description
 ------      -----------
 distance    length of the segment, in variable, human friendly units.
 polyline    a string encoding the points in the path.
 levels      a string containing information used for rendering the 
             polyline in an application like Google Maps.
 panel       HTML+JavaScript version of the driving directions for
             use in an AJAX application.  Google Maps uses this
             data in the left hand panel of a directions search.
 segments    a list of Geo::Google::Segment segments along the path.
             a segment has 0..1 driving directions associated with it.
 time        a time estimate, in variable, human-friendly units for how long
             the segment will take to travel by automobile.
 locations   an array of Geo::Google::Location objects containing 
             the start point (element 0 of the array), the final
             destination (the last element of the array), and any 
             waypoints between them that were used in the directions
             query.  There will always be at least two elements in
             this array (start point and final destination).  
 toXML       a method that renders the path in XML that could be used as 
             part of a Google Earth KML file.
 toJSON      a method that renders the path in JSON that could be used 
             with Google Maps.

=head1 SEE ALSO

L<Geo::Google>

=head1 AUTHOR

Allen Day E<lt>allenday@ucla.eduE<gt>, Michael Trowbridge 
E<lt>michael.a.trowbridge@gmail.comE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (c) 2004-2007 Allen Day.  All rights
reserved. This program is free software; you can redistribute it
and/or modify it under the same terms as Perl itself.

=cut
